stonesoup/measures.py

```diff
@@ -401,3 +401,63 @@ def __call__(self, state1, state2):
         mins = [min(s1, s2) for s1, s2 in zip(s1, s2)]
         maxs = [max(s1, s2) for s1, s2 in zip(s1, s2)]
         return np.sum(mins)/np.sum(maxs)
+
+
+class KLDivergence(Measure):
+    r"""Kullback-Leibler divergence between two distributions
+
+    Kullback-Leibler divergence, also referred to as relative entropy, is a
+    statistical distance. It describes how a probability distribution is
+    different from another. The expression for Kullback-Leibler divergence
+    is given by [1]_
+
+    .. math::
+        D_{KL}(P\Vert Q) = \sum_x P(x)\log \frac{P(x)}{Q(x)},
+
+    where :math:`P(x)` is the first distribution, or ``state1`` and :math:`Q(x)`
+    is the second distribution or, ``state2``. It is worth noting that Kullback-Leibler
+    divergence is not symmetric under interchange of :math:`P(x)` and :math:`Q(x)`. The
+    implementation here uses natural log meaning the returned divergence has units in nats.
+    This implementation assumes a discrete probability space and therefore only accepts
+    :class:`~.ParticleState`.
+
+    References
+    ----------
+    .. [1] MacKay, David J. C. 2003. Information Theory, Inference and Learning
+       Algorithms, 1st Ed. Cambridge University Press, """
+
+    def __call__(self, state1, state2):
+        r"""
+        Computes the Kullback–Leibler divergence from ``state1`` to ``state2``
+
+        Parameters
+        ----------
+        state1 : :class:`~.ParticleState`
+        state2 : :class:`~.ParticleState`
+
+        Returns
+        -------
+        float
+            Kullback–Leibler divergence from ``state1`` to ``state2``
+
+        """
+        if isinstance(state1, ParticleState) and isinstance(state2, ParticleState):
+            if len(state1.particles) == len(state2.particles):
+
+                log_term = np.zeros(state1.log_weight.shape)
+
+                invalid_indx = (np.isinf(state1.log_weight) | np.isnan(state1.log_weight)
+                                | np.isinf(state2.log_weight) | np.isnan(state2.log_weight))
+
+                # Do not consider NANs and inf in the subtraction
+                log_term[~invalid_indx] = state1.log_weight[~invalid_indx] \
+                    - state2.log_weight[~invalid_indx]
+
+                kld = np.sum(np.exp(state1.log_weight)*log_term)
+            else:
+                raise ValueError(f'The input sizes are not compatible '
+                                 f'({len(state1.particles)} != {len(state2.particles)})')
+        else:
+            raise TypeError('state1 or state2 is not a ParticleState')
```

Comment:
Do you think this can be extended to other states in the future? Maybe this could raise a `NotImplementedError`, stating that only `ParticleState` is currently supported?

Reply:
The only reason it only supports particle states currently is due to this being the discretised version of KLD. You could use it with any continuous distribution if we implemented a sampling technique to discretise the provided continuous distribution, perhaps using the sampler. I am happy to change this to a `NotImplementedError` for now.

Reply:
I will also add this to the `ExpectedKLDivergence` and `MultiUpdateExpectedKLDivergence` `__init__` methods as a similar particle constraint is enforced there and they can the problems relating to continuous states there can also be overcome in this way.